### PR TITLE
MQTT implementation

### PR DESCRIPTION
**Pull request by vallettea**
This is a start.
The code is not neat and things should be refactored a bit in some places, but a least we get a feeling of what if tends to be.
TODO
- [ ] dashboard should subscribe to `measurement`
- [ ] admin backend should subscribe to `command` and `status`
- [ ] socket.io should propagate message to frontends

### CODE REVIEW

File: broker/index.js [hunk 1]
```diff
@@ -34,323 +36,157 @@ var debug = function() {
...
+                    'start_hour': CONF.start_hour,
+                    'stop_hour': CONF.stop_hour
                 })
-                .then(function(sensor){
-                    console.log('sending config');
-                    var date = new Date();
-                    // Send config to the sensor
-                    sendCommand(tcpSocketSensor, 'init '+ [sensor.data_period, sensor.start_time, sensor.stop_time, date.toISOString()].join(" "));
+                .then(function(createdSensor){
+                    return createdSensor;
```
Comment (iOiurson) [inline broker/index.js]: I think there is no need for that, the sensor is already returned.

File: broker/index.js [hunk 2]
```diff
@@ -34,323 +36,157 @@ var debug = function() {
...
-                    else { 
-                        return sensors[0];
-                    }
-                    
+var authenticate = function(client, username, token, callback) {
+    var authorized = (token.toString() === PRIVATE.token);
+    if (authorized) {
+
+        // check if sensor exists in db
+        database.Sensors.get(client.id)
```
Comment (iOiurson) [inline broker/index.js]: Shouldn't we call `client.id` -> `client.sim`for the sake of clarity ?? I guess that the `client` should send its `id` under `sim`.

File: broker/index.js [hunk 3]
```diff
@@ -34,323 +36,157 @@ var debug = function() {
...
-
-
-
-
-
-
-// ############### API communication ######################
-
-var tcpServerToAdminApp = net.createServer(function(tcpSocketAdminApp) {
+    callback(null, authorized);
```
Comment (iOiurson) [inline broker/index.js]: Why is this not in a `.then()` ?

File: broker/index.js [hunk 4]
```diff
@@ -34,323 +36,157 @@ var debug = function() {
...
+          
                 break;
-
-            case 'message':
-                if (data.message.decoded === 'init') {
-                    var date = new Date();
-                    sendCommand(socket, 'date ' + date.toISOString())
+            case "status":
+                switch (subtopics[1]) {
+                    case "unitialized":
```
Comment (iOiurson) [inline broker/index.js]: Careful: `uninitialized`, not `unitialized`

File: broker/index.js [hunk 5]
```diff
@@ -34,323 +36,157 @@ var debug = function() {
...
-                else
-                    console.log('sim2socket[antSim] undefined !')
-            });
 
+    mqttServer.on('published', function(packet, client) {
+        console.log("MQTT ", packet.topic, packet.payload.toString())
+        var subtopics = packet.topic.split("/");
+        var id;
+        if(id2sensor[subtopics[0]]) {
+            id = subtopics[0];
```
Comment (iOiurson) [inline broker/index.js]: What if `sensor` is not known ??

Comment (vallettea) [inline broker/index.js]: it's not possible because an unknown sensor cannot publish

File: broker/index.js (hunk 2 above)
Comment (vallettea) [inline broker/index.js]: yes

File: broker/index.js (hunk 1 above)
Comment (vallettea) [inline broker/index.js]: yes

File: broker/index.js (hunk 3 above)
Comment (vallettea) [inline broker/index.js]: I don't know, i took this auth function from the doc. maybe we should adapt it

File: broker/index.js (hunk 4 above)
Comment (vallettea) [inline broker/index.js]: :cry:

File: broker/index.js (hunk 2 above)
Comment (iOiurson) [inline broker/index.js]: `client.id` is compulsory, so let's stick with it.

File: broker/index.js (hunk 3 above)
Comment (iOiurson) [inline broker/index.js]: There was the mother tickling problem :sob: ...
`callback` should be called after the resolution of the promise.

**Pull request merged**